quesma/clickhouse/clickhouse.go

```diff
@@ -518,42 +518,79 @@ func (lm *LogManager) generateNewColumns(
 	return alterCmd
 }
 
-func generateNonSchemaFieldsString(attrsMap map[string][]interface{}) (string, error) {
+// This struct contains the information about the columns that aren't part of the schema
+// and will go into attributes map
+type NonSchemaField struct {
+	Key   string
+	Value string
+	Type  string // inferred from incoming json
+}
+
+func convertNonSchemaFieldsToString(nonSchemaFields []NonSchemaField) string {
+	if len(nonSchemaFields) <= 0 {
+		return ""
+	}
+	attributesColumns := []string{AttributesValuesColumn, AttributesMetadataColumn}
 	var nonSchemaStr string
+	for columnIndex, column := range attributesColumns {
+		var value string
+		if columnIndex > 0 {
+			nonSchemaStr += ","
+		}
+		nonSchemaStr += "\"" + column + "\":{"
+		for i := 0; i < len(nonSchemaFields); i++ {
+			if columnIndex > 0 {
+				value = nonSchemaFields[i].Type
+			} else {
+				value = nonSchemaFields[i].Value
+			}
+			if i > 0 {
+				nonSchemaStr += ","
+			}
+			nonSchemaStr += fmt.Sprintf("\"%s\":\"%s\"", nonSchemaFields[i].Key, value)
+		}
+		nonSchemaStr = nonSchemaStr + "}"
```

Comment:
Hmm would StringBuilder make this look a little better? I'm nor sure myself, I'll leave that up to you :)

Reply:
Yes, good point